src/unread/unreadHuddlesReducer.js

```diff
@@ -18,20 +18,22 @@ import { NULL_ARRAY } from '../nullObjects';
 const initialState: UnreadHuddlesState = NULL_ARRAY;
 
 const eventNewMessage = (state, action) => {
-  if (action.message.type !== 'private') {
+  const { message } = action;
+
+  if (message.type !== 'private') {
```

Comment:
Now, after the early return on `message.type !== 'private'`, Flow
    will remember that `message` is a PM, even after some intervening
    things happen, like function calls. This will help when
    `pmUnreadsKeyFromMessage` starts taking a `PmMessage` soon.

Yeah. Without this change, it's not so much that Flow doesn't *remember* the type as it is that Flow doesn't *trust* that it will still have that type. The basic issue is that an intervening function call (like to `recipientsOfPrivateMessage`) could go and mutate `action`, causing `action.message` to be some completely other object from the object it was.

Here's the docs on this behavior:
https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations

Reply:
Looking again, I think you're right.

I think initially I wanted to say something like that ('trust' instead of 'remember), but then I thought something like, "well, if that's all that's happening, and in the old code Flow didn't want to trust that `action` wasn't being mutated, then why is Flow happy in the new code to trust that `message` isn't being mutated to have a different `message.type`?" And I didn't investigate further.

I think the answer to that is that `PmMessage`, the type that `message` is refined to after the early return, is `$ReadOnly`.

Which I think suggests that we might have taken a different angle than the one in this commit: we might have marked `EventNewMessageAction` as `$ReadOnly`. (Untested.) We should do that anyway, I think, but not urgently, and this change should be fine to make.

Reply:
> I think initially I wanted to say something like that ('trust' instead of 'remember'), but then I thought something like, "well, if that's all that's happening, and in the old code Flow didn't want to trust that `action` wasn't being mutated, then why is Flow happy in the new code to trust that `message` isn't being mutated to have a different `message.type`?" And I didn't investigate further.

That is indeed a good question. The answer isn't awesome.


> I think the answer to that is that `PmMessage`, the type that `message` is refined to after the early return, is `$ReadOnly`.

Well, the invalidation happens even if the function in question isn't passed any arguments -- which is appropriate, because the object we're looking at may be known by another name somewhere else (e.g., it could be a global somewhere), and so the function we call might happen to somehow know about it and get to it.

And if that callee knows about the object by another name, it might know it by a non-`$ReadOnly` type. When we have a read-only type, that only reflects a promise this code is *making* to any other code that passes it the value; it doesn't come with any promises in the other direction, that the code that supplies the value won't be keeping a reference that it might later allow to be used to mutate it.

So that explanation doesn't seem to be the story.


What the behavior actually seems to be is something like: if you have an object type with some refinements on its properties, and you call a function, then the refinements get invalidated. But if the top-level type is a union and you have some refinements on it, then calling a function doesn't invalidate that.

In particular, if you have `x: { t: 'n' } | { t: 's' }` and a refinement `t !== 'n'`, then that will persist across a function call. But if you have `x: { t: 'n' | 's' }` -- which sure looks awfully similar -- and the same refinement, then that will not persist. [Here's a Try Flow demo](https://flow.org/try/#0C4TwDgpgBAKgFgSwHYHMoF4oG8rAFxQDkShANFAIYFICuAtgEYQBOAvlAD7a4GEDOZKAwJ9gzZGlYBuAFAyAZjSQBjYAgD2SKPIAUADwLwJASmwyoFqAnlR9AOmAZ0mYoVPMIwGsySzLUAHoAqAAlCHlkCDoIJGA+SjAwABsQO3NLe3wiAWM-DL07YShRcVRc9IsUHXL-INDwyOjY+NEEJKSE5JByYDh1GhQ4KAADFGGoZX6kgBMoOAoAN2g6GmAKYAhZ4b1htP9M3hy8i3sikpNZVjlFFTVNbV09A25no1QoVlMsCqsbfQKCo5nC4SO5PN5fD86mEIkgojE4p0Unt8gCHIc3MdbE87AUzmILj8qjVLNCGnCmoikOooElNCgWEjUhYAILyDbMSjtcijcaTGgzOaLCBQ4IrNYbLZPcZ8Gm9dYjHE7IQQZQUaLFdTRXoSKAQJJ8CAok5K9HZTE-f64woiAllKT+GRXBRKVQaLRVZ50BB6TZfaRyGR1ADKBLAUH6jl60AgejEFCgYGY6kgzDUED4DtE6mYICg03UGZIjmU81QRrkoEgsEQqAATBhuFlXB9OE2MR9rq67lp5HX9IZayg618ftZsQ4nCC3FAPF4fFiybD4c0mcaJ82jkTqovgjDGgiWmp2kyen0BkNeRMprN5ksoOL1ptFbtLYCMeVnTc3fd5H3-gQOCvEODafGY-jjlagJTkQoKzuCC6ivUy6UvEFCJMib5opuFr+MSu7IQeq7UrS9KMuhXR2Ky7LkdyIxjNeAq3sKSGPpKirKrKuDzI42wAuMTBqhqsrakOeoGhW+ymjhn6BnUACy6j3tGUBKO6VhIHwCDTNAKnqAwABWqpRuAFZ1AAcuoADu3HQNmub5oW8SlhQ5YAISVqZNYSAAzI2ODNiQbb8IQraBMEdLqAA1mhpGOEkCBRdAbzDuQBZFoQjgIMAAD8Xa3Op-7Ab5o4QX80HArBM5zhCBH7hSh5rm+ZohSSlQ7kh9UroiABUJF0uWzA9U1tTBGyHJckkPIMfygp3ssqxPlKr5SS1W5ft2hVFc8QGDr5HylZYkHSTBrhgvOkKjYRDWrhRmFSdhH5YvhnXkt18T9WRnJ3cyUDjbRU30XyN5CksrGLexfEynKPEcbiAmquqdlap4Yn6oa65QWtFpXEAA).

This leaves the behavior not completely sound in the face of mutation. I suspect their reasoning here reflects a balance: with the way it does behave, you can always make the refinement stick by just picking out the relevant property as a variable, perhaps a property-of-property, however deep the type goes as object types before you reach a union. That's straightforward to do and not messy in the code, and it does really protect you against possible mutation of those objects (even if that isn't a real likely sort of bug, which it isn't in our codebase.) But if they refused more thoroughly to keep refinements across function calls, there isn't a great way we could adjust the code to respond to that. A lot of code would have to become messier, not worth the exclusion of this fairly unlikely class of bug.

Reply:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Flow.20refinements/near/1262824

Reply:
I think Flow is being consistent here.

* It’s justified in assuming that the `t` property of `x: { t: 'n' } | { t: 's' }` won’t be mutated by external functions, because mutating it would be a type error in those functions.
* The difference with `xx: { x: { t: 'n' } | { t: 's' } }` is *not* that the `t` property of `xx.x` might be mutated, but rather that the `x` property of `xx` might be mutated.
* The difference with `x: { t: 'n' | 's' }` is that mutating *this* `t` property would not be a type error.

[Demo](https://flow.org/try/#0DYUwLgBA5hC8EAoCUcB8EDeBfA3AKDzAE8AHECAFQAsBLAOygCY5MIwAuCAcjq4iwgAfVh24BnPlgIAzAK50AxmBoB7OhGmMEAD07V6TFBjwRTEGtMTaAdJFj3uvFACdws53XxmrtzlwlIXmZQyEGmOr7iXIEQAPSxECoA1nhSeApqYpDajHq0DMzwGGx+Evz4MPDIaJgmZjm2LDxcOHEJIM7OKs4AhKn4mjqMgTLySqrq0oPauqyz+gX8RnWmFlY2NnYOzS5uHmHr1pul0QchI946G5H+p20QHV29qQQZdFkQM7lzeQbMAkVPpxiqJbvxynhKogULB0MZvF8jiwQX5eOV7sl+ngploviNCKRyAsoABmZElRx8YRgtJyRTKNQaXG-BgkgA00E41VhEAAbioaAATZbeNYRLbwHYQVxgdyeFY+UEBM6hBXik4xeIPTrdPppN4fbQklmk8lKyQVFjcuEKo2NSW8VpazG0sYMybTWbFeb5U1YEVmMUzI727ZOaV7eWXYPHKIXYKq6PXc2a9o6576zLZI3AoGUX1kgE-EQncG4SFWmE2hF2zZNR0YlI1klIwGgtG4RtYoA).

Reply:
> * It’s justified in assuming that the `t` property of `x: { t: 'n' } | { t: 's' }` won’t be mutated by external functions, because mutating it would be a type error in those functions.

Hmm, thanks! I hadn't expected that that would be a type error, but on studying this further, I agree. More in that chat thread.

One observation that helped make sense of this for me is:
```js
(x: {t: 'n'} | {t: 's'}): {+t: 'n' | 's'} => x; // ok
(x: {t: 'n' | 's'}): {+t: 'n'} | {+t: 's'} => x; // error!
```
So if you have a `{+t: 'n'} | {+t: 's'}`, it can't have come from something that had a `{t: 'n' | 's'}`, and was thereby holding on to an option to mutate `t` from 'n' to 's' and back.